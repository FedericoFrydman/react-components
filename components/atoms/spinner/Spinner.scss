@import '../../assets/styles/colors.scss';

$spinner-border-size: 0.2rem;

.lsux-spinner-blocker {
  align-items: center;
  background-color: rgba(0, 0, 0, 0.4);
  display: flex;
  height: 100%;
  justify-content: center;
  left: 0;
  overflow: auto;
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 1002;
}

.lsux-spinner {
  border-radius: 50%;
  height: 50px;
  margin: 20px;
  position: relative;
  width: 50px;

  &__center {
    -webkit-animation: rotate-spinner 1s infinite linear; /* Safari */
    animation: rotate-spinner 1s infinite linear;
    border-radius: 50%;
    border-right: $spinner-border-size solid rgba(255, 255, 255, 0.6);
    height: 100%;
    width: 100%;

    &--neutral {
      border-bottom: $spinner-border-size solid $P200;
      border-left: $spinner-border-size solid $P200;
      border-top: $spinner-border-size solid $P200;
    }

    &--brand {
      border-bottom: $spinner-border-size solid var(--brand-color-prim200);
      border-left: $spinner-border-size solid var(--brand-color-prim200);
      border-top: $spinner-border-size solid var(--brand-color-prim200);
    }
  }

  &--small {
    height: 16px;
    width: 16px;
  }

  &--medium {
    height: 24px;
    width: 24px;
  }

  &--large {
    height: 48px;
    width: 48px;
  }

  &--xlarge {
    height: 96px;
    width: 96px;
  }

  /* Safari */
  @-webkit-keyframes rotate-spinner {
    from {
      transform: rotate(0);
    }
    to {
      transform: rotate(360deg);
    }
  }

  @keyframes rotate-spinner {
    from {
      transform: rotate(0);
    }
    to {
      transform: rotate(360deg);
    }
  }
}
