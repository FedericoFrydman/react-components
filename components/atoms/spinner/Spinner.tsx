import { forwardRef, HTMLProps, Ref } from 'react';
import cn from 'classnames';

import './Spinner.scss';

export type SpinnerSize = 'small' | 'medium' | 'large' | 'xlarge';
export type SpinnerColor = 'neutral' | 'brand';

export interface SpinnerProps extends Omit<HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Sets the size of the spinner
   * @default medium
   */
  spinnerSize?: SpinnerSize;
  /**
   * Spinner color
   * @default neutral
   */
  color?: SpinnerColor;
  /**
   * Blocks the screen
   * @default false
   */
  blocking?: boolean;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const SpinnerComponent = (
  {
    blocking = false,
    spinnerSize = blocking ? 'large' : 'medium',
    color = 'neutral',
    classNames = [],
    ...props
  }: SpinnerProps,
  ref: SpinnerProps['ref'],
) => {
  const classSpinner = 'lsux-spinner';
  const cnSpinner = cn(
    {
      [`${classSpinner}`]: true,
      [`${classSpinner}--${spinnerSize}`]: spinnerSize,
      [`${classSpinner}--${color}`]: color,
      [`spin`]: true,
    },
    classNames,
  );

  const classSpinnerCenter = `${classSpinner}__center`;
  const cnCircle = cn({
    [classSpinnerCenter]: true,
    [`${classSpinnerCenter}--${color}`]: true,
  });

  const spinnerEl = (
    <div {...props} className={cnSpinner} ref={ref}>
      <div className={cnCircle}></div>
    </div>
  );

  if (blocking) {
    return <div className="lsux-spinner-blocker">{spinnerEl}</div>;
  } else {
    return spinnerEl;
  }
};

/**
 * Use our circular spinner to show users a task is loading or in progress, and will take an indeterminate amount of time. There should only be a single spinner on a page at one time.
 */

export const Spinner = forwardRef<HTMLDivElement, SpinnerProps>(SpinnerComponent);
