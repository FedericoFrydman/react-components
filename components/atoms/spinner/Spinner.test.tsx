import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Spinner, SpinnerColor, SpinnerProps, SpinnerSize } from './Spinner';

describe('Spinner Component', () => {
  const baseClass = `lsux-spinner`;
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Spinner />);
    testId = 'test-spinner';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the text the provided classNames', () => {
          const text = renderSpinner({ classNames });

          expect(text).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('spinnerSize', () => {
      const spinnerSizes: SpinnerSize[] = ['small', 'medium', 'large', 'xlarge'];
      let spinnerSizeClass: string;

      describe('when "spinnerSize" is not defined', () => {
        it('defaults "spinnerSize" to the "medium"', () => {
          const spinnerEl = renderSpinner({ spinnerSize: 'medium' });

          expect(spinnerEl).toHaveClass(`${baseClass}--medium`);
        });

        function itAssignsTheRightClassForSpinnerSize(spinnerSize: SpinnerSize) {
          describe(`when "spinnerSize" is "${spinnerSize}"`, () => {
            beforeEach(() => {
              spinnerSizeClass = `${baseClass}--${spinnerSize}`;
            });

            it(`assigns the element the ${baseClass}--${spinnerSize} class`, () => {
              const spinnerEl = renderSpinner({ spinnerSize });

              expect(spinnerEl).toHaveClass(spinnerSizeClass);
            });
          });
        }

        spinnerSizes.forEach((spinnerSize) => itAssignsTheRightClassForSpinnerSize(spinnerSize));
      });
    });

    describe('color', () => {
      const colors: SpinnerColor[] = ['brand', 'neutral'];
      let spinnerColorClass: string;

      describe('when "color" is not defined', () => {
        it('defaults "color" to the "neutral"', () => {
          const spinnerEl = renderSpinner();

          expect(spinnerEl).toHaveClass(`${baseClass}--neutral`);
        });
      });

      function itAssignsTheRightClassForSpinnerColor(color: SpinnerColor) {
        describe(`when "color" is "${color}"`, () => {
          beforeEach(() => {
            spinnerColorClass = `${baseClass}--${color}`;
          });

          it(`assigns the element the ${baseClass}--${color} class`, () => {
            const spinnerEl = renderSpinner({ color });

            expect(spinnerEl).toHaveClass(spinnerColorClass);
          });
        });
      }

      colors.forEach((color) => itAssignsTheRightClassForSpinnerColor(color));
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the spinner', () => {
      expect(renderSpinner({ ref })).toEqual(ref.current);
    });
  });

  const renderSpinner = (props: SpinnerProps = {}): HTMLElement => {
    render(<Spinner {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
