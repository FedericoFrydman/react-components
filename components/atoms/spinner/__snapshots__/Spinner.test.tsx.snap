// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Component should render the component 1`] = `
<div
  className="lsux-spinner lsux-spinner--medium lsux-spinner--neutral spin"
>
  <div
    className="lsux-spinner__center lsux-spinner__center--neutral"
  />
</div>
`;
